include/nbl/builtin/hlsl/bxdf/fresnel.hlsl

```diff
@@ -17,131 +18,354 @@ namespace hlsl
 namespace bxdf
 {
 
-namespace impl
+namespace fresnel
 {
-template<typename T>
-struct orientedEtas;
 
-template<>
-struct orientedEtas<float>
+template<typename T NBL_PRIMARY_REQUIRES(is_scalar_v<T> || is_vector_v<T>)
+struct OrientedEtas
 {
-    static bool __call(NBL_REF_ARG(float) orientedEta, NBL_REF_ARG(float) rcpOrientedEta, float NdotI, float eta)
+    using scalar_type = typename vector_traits<T>::scalar_type;
+
+    static OrientedEtas<T> create(scalar_type NdotI, T eta)
     {
-        const bool backside = NdotI < 0.0;
-        const float rcpEta = 1.0 / eta;
-        orientedEta = backside ? rcpEta : eta;
-        rcpOrientedEta = backside ? eta : rcpEta;
-        return backside;
+        OrientedEtas<T> retval;
+        retval.backside = NdotI < scalar_type(0.0);
+        const T rcpEta = hlsl::promote<T>(1.0) / eta;
+        retval.value = retval.backside ? rcpEta : eta;
+        retval.rcp = retval.backside ? eta : rcpEta;
+        return retval;
     }
+
+    T value;
+    T rcp;
+    bool backside;
```

Comment:
BxDFs need to track backsides, not the oriented Eta 

Oriented Eta after being computed is oblivious to originally being a backside or not

https://github.com/Devsh-Graphics-Programming/Nabla/pull/811#discussion_r2046786460